Initial implementation of a Spark connector

n-riesco opened the pull request.

n-riesco commented on test/backend/datastores.livy.spec.js:
```diff
@@ -76,7 +76,8 @@ df.registerTempTable("ALCOHOL_CONSUMPTION_BY_COUNTRY_2010")
 
         return query(`SELECT * FROM ${tableName}\nLIMIT 5`, connection).then(results => {
-            assert.deepEqual(results.rows, [
+            const roundedToSinglePrecision = results.rows.map(r => [r[0], Number(r[1].toPrecision(6))]);
```
It looks like when the table was created in Spark, the source CSV file was imported as single-precision numbers. But when we query Spark/Livy, the results are reported as double-precision numbers.

@chriddyp Have you come across this issue with any of the other connectors?

chriddyp replied:
Hm, I haven't come across this although I haven't really investigated too deeply either

n-riesco replied:
@tarzzz @chriddyp Here's an untested explanation: I've just noticed that [the script to create the sample Hive table declares the column as `FLOAT`](https://github.com/plotly/plotly-database-connector/blob/2eae53de7347051328ea3cd6bd428609d765a90d/test/docker/hadoop-spark/create_hive_tables.py#L14).

tarzzz commented:
@n-riesco @chriddyp  .. 

My only concern is with the workaround for float-precision in tests, but we can revert it later once test database is fixed (i.e. using decimal instead of float in its column)..   Good for now.. 💃

chriddyp commented:
It looks like the build failed:
```
-----> Build failed
 !     Outdated Yarn lockfile
       Your application contains a Yarn lockfile (yarn.lock) which does not
       match the dependencies in package.json. This can happen if you use npm
       to install or update a dependency instead of Yarn.
       Please run the following command in your application directory and check
       in the new yarn.lock file:
       $ yarn install
       $ git add yarn.lock
       $ git commit -m "Updated Yarn lockfile"
       $ git push heroku master
    
       https://kb.heroku.com/why-is-my-node-js-build-failing-because-of-an-outdated-yarn-lockfile
 !     Push rejected, failed to compile Node.js app.
 !     Push failed
```

chriddyp commented on backend/persistent/datastores/livy.js:
```diff
@@ -0,0 +1,208 @@
+        `val r = plotlyContext.sql("show tables").collect\n%json r`;
+    return sendRequest(connection, code)
+        .then(json => {
+            if (!Array.isArray(json) || json.length <= 0) {
```
why `<`? can this just be `=== 0`?

n-riesco replied:
It's just a style thing. I'll change it to `===`.

chriddyp commented on backend/persistent/datastores/livy.js:
```diff
@@ -0,0 +1,208 @@
+    return getState(connection).then(state => {
+        return (state === 'idle') ?
+            connection :
+            wait(1000).then(function() {
```
😻  love this `wait`

chriddyp commented on sample-storage/connections.yaml:
```diff
@@ -78,3 +78,11 @@
+    timeout: 180
+    database: plotly
+    port: 8998
+    host: 104.155.170.7
```
should there be another one in here for the sample IBM DB? 

It's not really documented, but `sample-storage` is used for the test heroku instance (heroku doesn't have a persistent file system, so we need to commit these credentials as part of the repo and then point the settings folder to this `sample-storage` folder to preload the available connections).

chriddyp replied:
NVM, it's already there :woman_facepalming:

chriddyp commented:
Looks good! Just a couple of small comments + the build, then 💃

tarzzz commented:
@chriddyp The build should succeed after merging #175 and rebase, since it is related to yarn lockfile..

chriddyp commented:
OK, I'll create a review app there. It looks like the review apps aren't being created automatically since this PR is based off of a fork.
Here's where you can create the review apps in heroku:
![image](https://user-images.githubusercontent.com/1280389/30133238-28fd4e98-9321-11e7-81d0-be1a896733e0.png)

chriddyp commented:
Could we add Spark and DB2 to the end of the list but right before Apache Drill? Right now, they are the first two items but they are probably less common than MySQL
![image](https://user-images.githubusercontent.com/1280389/30133364-8ff6f248-9321-11e7-9555-5b0e1cf0c42d.png)
but more common than Apache Drill

chriddyp commented:
The Spark Instance "Had trouble connecting" but I'm assuming that this is an unrelated issue?
![image](https://user-images.githubusercontent.com/1280389/30133415-c95e49d2-9321-11e7-88da-b4b21409054d.png)

n-riesco commented:
Hi @chriddyp I've addressed the issues you raised. I've also added a commit to improve error handling from Spark.

The PR is ready for review again.

chriddyp commented:
Great, thanks @n-riesco ! 💃  from me. @tarzzz ?

tarzzz commented:
Yep, LGTM.. 💃 🎉 🎉
